File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/controllers/commands/get_focus_info.py
# nvdaMcpBridge domain -- GetFocusInfoHandler: answer "where am I".
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: command handler for `getFocusInfo`.
# A null focus yields an empty FocusInfoResult, not an error.

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .... import protocol
from .command_handler import CommandHandler

if TYPE_CHECKING:
	from .session_context import SessionContext


class GetFocusInfoHandler(CommandHandler):
	def execute(self, ctx: SessionContext, request: protocol.Request) -> Any:
		info = ctx.adapter_set.focus_inspector.focus_info()
		return protocol.FocusInfoResult(
			name=info.name,
			role=info.role,
			states=info.states,
			value=info.value,
			appModule=info.app_module,
		)

File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Controllers/Commands/GetFocusInfo.swift
// ROLE: controller -- `getFocusInfo`: answer "where am I".
// BUILT BY: Registry. DRIVES: the FocusInspector port, and nothing else.
// It never requests a permission; the adapter decides how rich an answer to give from a read
// that shows no dialog.

import ScreenReaderWire

public final class GetFocusInfoHandler: CommandHandler {
	public init() {}

	public func execute(_ context: SessionContext, _ request: Request) throws -> any Encodable {
		let adapters = try readerEdge(context)
		let focus: FocusSnapshot
		do {
			focus = try adapters.focusInspector.focusInfo()
		} catch let failure as FocusError {
			throw CommandError("the focus could not be read: \(failure.description)")
		}
		return FocusInfoResult(
			name: focus.name,
			role: focus.role,
			states: focus.states,
			value: focus.value,
			appModule: focus.appModule
		)
	}

	private func readerEdge(_ context: SessionContext) throws -> AdapterSet {
		guard let adapters = context.adapters else {
			throw CommandError("focus was read before `hello` built the reader edge")
		}
		return adapters
	}
}
